base/iobuffer.jl

```diff
@@ -164,15 +268,43 @@ PipeBuffer(maxsize::Integer) = (x = PipeBuffer(StringMemory(maxsize), maxsize =
 _similar_data(b::GenericIOBuffer, len::Int) = similar(b.data, len)
 _similar_data(b::IOBuffer, len::Int) = StringMemory(len)
 
-function copy(b::GenericIOBuffer)
-    ret = typeof(b)(b.reinit ? _similar_data(b, 0) : b.writable ?
-                    copyto!(_similar_data(b, length(b.data)), b.data) : b.data,
-                    b.readable, b.writable, b.seekable, b.append, b.maxsize)
-    ret.size = b.size
-    ret.ptr  = b.ptr
-    ret.mark = b.mark
-    ret.offset = b.offset
-    return ret
+# Note: Copying may change the value of the position (and mark) for un-seekable streams.
+# However, these values are not stable anyway due to compaction.
+
+function copy(b::GenericIOBuffer{T}) where T
+    if b.reinit
+        # If buffer is used up, allocate a new size-zero buffer
+        # Reinit implies wriable, and that ptr, size, offset and mark are already the default values
+        return typeof(b)(_similar_data(b, 0), b.readable, b.writable, b.seekable, b.append, b.maxsize)
+    elseif b.writable
+        # Else, we just copy the reachable bytes. If buffer is seekable, all bytes
+        # after offset are reachable, since they can be seeked to
+        used_span = if b.seekable
+            b.offset + 1 : b.size
+        else
+            # Even non-seekable streams can be seeked using `reset`. Therefore, we need to
+            # copy all data from mark if it's set and below ptr.
+            (b.mark > -1 ? min(b.ptr, b.mark) : b.ptr) : b.size
+        end
+        len = length(used_span)
+        data = copyto!(_similar_data(b, len), view(b.data, used_span))
+        ret = typeof(b)(data, b.readable, b.writable, b.seekable, b.append, b.maxsize)
+        ret.size = len
+        ret.offset = 0
+        ret.ptr = b.ptr - first(used_span) + 1
+        ret.mark = b.mark < 0 ? -1 : (b.mark - first(used_span) + 1)
+        return ret
+    else
+        # When the buffer is just readable, they can share the same data, so we just make
+        # a shallow copy of the IOBuffer struct.
+        # Use unsafe method because we want to allow b.maxsize to be larger than data, in case that
```

Comment:
The wording has to be updated now, maybe like this:

```suggestion
        # Use internal constructor because we want to allow b.maxsize to be larger than data, in case that
```

Reply:
Done